### PR TITLE
[Feature] Support MixMIM Pretrain and Finetuning

### PR DESCRIPTION
**Pull request by wangbo-zhao**
Add MixMIM [MixMIM: Mixed and Masked Image Modeling for Efficient Visual Representation Learning](https://arxiv.org/abs/2205.13137) to mmselfsup.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here and update the documentation.

## Checklist

**Before PR**:

- [ ] Pre-commit or other linting tools are used to fix the potential lint issues.
- [ ] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [x] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects, like MMDet or MMSeg.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

**Comment by wangbo-zhao** [comment]
MixMIMTransformer in mmselfsup/models/backbones/mixmim_backbone.py can be removed, when the PR in MMCls is merged.

**Comment by wangbo-zhao** [comment]
Currently，mmcls do not contains 'mmcls::_base_/models/mixmim/mixmim_base.py', which leads to the fail in pr_stage_test

**Comment by wangbo-zhao** [comment]
yiI23eRsVp

File: README.md [hunk 1]
```diff
@@ -137,6 +137,7 @@ Supported algorithms:
 - [x] [CAE (arXiv'2022)](https://github.com/open-mmlab/mmselfsup/tree/dev-1.x/configs/selfsup/cae)
 - [x] [MILAN (arXiv'2022)](https://github.com/open-mmlab/mmselfsup/tree/dev-1.x/configs/selfsup/milan)
 - [x] [BEiT v2 (arXiv'2022)](https://github.com/open-mmlab/mmselfsup/tree/dev-1.x/configs/selfsup/beitv2)
+- [x] [MixMIM (ArXiv'2022)](https://github.com/open-mmlab/mmselfsup/tree/dev-1.x/configs/selfsup/mixmim)
```
Comment (fangyixiao18) [inline README.md:141]: arXiv

File: README_zh-CN.md [hunk 2]
```diff
@@ -137,6 +137,7 @@ Useful Tools
 - [x] [CAE (arXiv'2022)](https://github.com/open-mmlab/mmselfsup/tree/dev-1.x/configs/selfsup/cae)
 - [x] [MILAN (arXiv'2022)](https://github.com/open-mmlab/mmselfsup/tree/dev-1.x/configs/selfsup/milan)
 - [x] [BEiT v2 (arXiv'2022)](https://github.com/open-mmlab/mmselfsup/tree/dev-1.x/configs/selfsup/beitv2)
+- [x] [MixMIM (ArXiv'2022)](https://github.com/open-mmlab/mmselfsup/tree/dev-1.x/configs/selfsup/mixmim)
```
Comment (fangyixiao18) [inline README_zh-CN.md:141]: arXiv

File: configs/selfsup/_base_/models/mixmim.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
+model = dict(
+    type='MixMIM',
+    data_preprocessor=dict(  # √
```
Comment (fangyixiao18) [inline configs/selfsup/_base_/models/mixmim.py]: the check mark can be removed

File: configs/selfsup/mixmim/metafile.yaml [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+      Training Resources: 16x A100-80G GPUs
+      Architecture:
+        - MixMIM ViT
+    Paper:
+        URL: https://arxiv.org/abs/2205.13137
+        Title: "MixMIM: Mixed and Masked Image Modeling for Efficient Visual Representation Learning"
+    README: configs/selfsup/mixmim/README.md
+
+Models:
+  - Name: mixmim-base-p16_16xb128-coslr-300e_1k
```
Comment (fangyixiao18) [inline configs/selfsup/mixmim/metafile.yaml]:
```suggestion
  - Name: mixmim-base-p16_16xb128-coslr-300e_in1k
```

File: configs/selfsup/mixmim/mixmim-base-p16_16xb128-coslr-300e_1k.py.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
+_base_ = [
```
Comment (fangyixiao18) [inline configs/selfsup/mixmim/mixmim-base-p16_16xb128-coslr-300e_1k.py.py]: update the name of the config file, it is supposed to be `mixmim-base-p16_16xb128-coslr-300e_in1k.py
`

File: configs/selfsup/mixmim/metafile.yaml [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    In Collection: MixMIM
+    Metadata:
+      Epochs: 300
+      Batch Size: 2048
+    Results: null
+    Config: configs/selfsup/mixmim/mixmim-base-p16_16xb128-coslr-300e_1k.py
```
Comment (fangyixiao18) [inline configs/selfsup/mixmim/metafile.yaml]:
```suggestion
    Config: configs/selfsup/mixmim/mixmim-base-p16_16xb128-coslr-300e_in1k.py
```

File: mmselfsup/models/algorithms/mixmim.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+from .base import BaseModel
+
+
+@MODELS.register_module()
+class MixMIM(BaseModel):
+    """MiXMIM. Implementation of `MixMIM: Mixed and Masked Image Modeling for
+    Efficient Visual Representation Learning.
+
+    <https://arxiv.org/abs/2205.13137>`_.
+    """
```
Comment (fangyixiao18) [inline mmselfsup/models/algorithms/mixmim.py:18]:
```suggestion
    """MiXMIM. 
    
    Implementation of `MixMIM: Mixed and Masked Image Modeling for
    Efficient Visual Representation Learning.
    <https://arxiv.org/abs/2205.13137>`_.
    """
```

File: mmselfsup/models/algorithms/mixmim.py [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+
+        Args:
+            inputs (List[torch.Tensor]): The input images.
+            data_samples (List[SelfSupDataSample]): All elements required
+                during the forward function.
+
+        Returns:
+            Dict[str, torch.Tensor]: A dictionary of loss components.
+        """
+        # ids_restore: the same as that in original repo, which is used
```
Comment (fangyixiao18) [inline mmselfsup/models/algorithms/mixmim.py]: remove unrelated comments

File: mmselfsup/models/backbones/mixmim_backbone.py [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+        use_checkpoint (bool): Whether use the checkpoint to
+        reduce GPU memory cost
+        range_mask_ratio (float): The range of mask ratio.
+            Defaults to 0.
+        init_cfg (dict, optional): Initialization config dict.
+            Defaults to None.
+    """
+
+    def __init__(self,
+                 arch='base',
```
Comment (fangyixiao18) [inline mmselfsup/models/backbones/mixmim_backbone.py]: please add typehint

File: mmselfsup/models/backbones/mixmim_backbone.py [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+                 patch_cfg=dict(),
+                 norm_cfg=dict(type='LN'),
+                 drop_rate=0.0,
+                 drop_path_rate=0.0,
+                 attn_drop_rate=0.0,
+                 use_checkpoint=False,
+                 range_mask_ratio=0.0,
+                 init_cfg: Optional[dict] = None) -> None:
+
+        super(MixMIMTransformerPretrain, self).__init__(
```
Comment (fangyixiao18) [inline mmselfsup/models/backbones/mixmim_backbone.py]:
```suggestion
        super().__init__(
```

File: mmselfsup/models/backbones/mixmim_backbone.py [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+        Args:
+            x (torch.Tensor): Image with data augmentation applied, which is
+                of shape B x L x C.
+            mask_ratio (float): The mask ratio of total patches.
+                Defaults to 0.5.
+
+        Returns:
+            Tuple[torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor]:
+                mask_s1, mask_s2, mask_s3, and mask.
+                  - mask_s1 (torch.Tensor): mask with stride of
```
Comment (fangyixiao18) [inline mmselfsup/models/backbones/mixmim_backbone.py]: please check the api in readthedocs here locally, it might have some displaying error

File: mmselfsup/models/backbones/mixmim_backbone.py [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+        This function generates mask and masks some patches randomly and get
+        the hidden features for visible patches.
+
+        Args:
+            x (torch.Tensor): Input images, which is of shape B x C x H x W.
+
+        Returns:
+            Tuple[torch.Tensor, torch.Tensor]:
+
+            Hidden features and mask.
```
Comment (fangyixiao18) [inline mmselfsup/models/backbones/mixmim_backbone.py]: also check api in readthedocs here

File: mmselfsup/models/heads/mixmim_head.py [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+        target = self.construct_target(target)
+
+        B, L, C = x_rec.shape
+
+        # unmix tokens
+        x1_rec = x_rec[:B // 2]
+        x2_rec = x_rec[B // 2:]
+
+        unmix_x_rec = x1_rec * mask + x2_rec.flip(0) * (1 - mask)
+        loss_rec = (unmix_x_rec - target)**2
```
Comment (fangyixiao18) [inline mmselfsup/models/heads/mixmim_head.py]: I think here can apply `PixelReconstructionLoss`, please check it

**Review by fangyixiao18 — CHANGES_REQUESTED** [review]

File: mmselfsup/models/heads/mixmim_head.py (hunk 13 above)
Comment (wangbo-zhao) [reply mmselfsup/models/heads/mixmim_head.py]: PixelReconstructionLoss  need a mask input, which may make it difficult to understand.

File: configs/selfsup/mixmim/mixmim-base-p16_16xb128-coslr-300e_in1k.py.py [hunk 14]
```diff
@@ -0,0 +1,46 @@
+_base_ = [
```
Comment (fangyixiao18) [inline configs/selfsup/mixmim/mixmim-base-p16_16xb128-coslr-300e_in1k.py.py]: redundant `.py` in filename

File: mmselfsup/models/algorithms/mixmim.py [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+from .base import BaseModel
+
+
+@MODELS.register_module()
+class MixMIM(BaseModel):
+    """MiXMIM.
+
+    Implementation of `MixMIM: Mixed and Masked Image Modeling for
+    Efficient Visual Representation Learning.
+
```
Comment (fangyixiao18) [inline mmselfsup/models/algorithms/mixmim.py]: please remove this blank line to avoid the displaying error in readthedocs

File: mmselfsup/models/backbones/mixmim_backbone.py [hunk 16]
```diff
@@ -0,0 +1,196 @@
...
+class MixMIMTransformerPretrain(MixMIMTransformer):
+    """MixMIM backbone during pretraining.
+    A PyTorch implement of : ` MixMIM: Mixed and Masked Image
+    Modeling for Efficient Visual Representation Learning
+    <https://arxiv.org/abs/2205.13137>`_
+    Args:
+        arch (str | dict): MixMIM architecture. If use string,
+            choose from 'base','large' and 'huge'.
+            If use dict, it should have below keys:
+            - **embed_dims** (int): The dimensions of embedding.
```
Comment (fangyixiao18) [inline mmselfsup/models/backbones/mixmim_backbone.py:27]: refine here according to the format in mmcls

File: mmselfsup/models/heads/mixmim_head.py (hunk 13 above)
Comment (fangyixiao18) [reply mmselfsup/models/heads/mixmim_head.py]: maybe add an `if-else` branch in the forward function to deal with the mask input, I think we can set the default value to `None` of the `mask` in the definition of the forward function. The current logic here is kind of different and not very appropriate, as we build loss according to the `dict` from config

**Review by fangyixiao18 — CHANGES_REQUESTED** [review]

File: mmselfsup/models/heads/mixmim_head.py (hunk 13 above)
Comment (wangbo-zhao) [reply mmselfsup/models/heads/mixmim_head.py]: good suggenstion. done

File: mmselfsup/models/losses/reconstruction_loss.py [hunk 17]
```diff
@@ -38,8 +38,10 @@ def __init__(self, criterion: str, channel: Optional[int] = None) -> None:
 
         self.channel = channel if channel is not None else 1
 
-    def forward(self, pred: torch.Tensor, target: torch.Tensor,
-                mask: torch.Tensor) -> torch.Tensor:
+    def forward(self,
+                pred: torch.Tensor,
+                target: torch.Tensor,
+                mask: torch.Tensor = None) -> torch.Tensor:
```
Comment (fangyixiao18) [inline mmselfsup/models/losses/reconstruction_loss.py]:
```suggestion
                mask: Optional[torch.Tensor] = None) -> torch.Tensor:
```

Comment (wangbo-zhao) [reply mmselfsup/models/losses/reconstruction_loss.py]: done

**Review by fangyixiao18 — APPROVED** [review]
LGTM

**Review by tonysy — APPROVED** [review]
LGTM
